torch/_dynamo/variables/misc.py

```diff
@@ -224,8 +224,17 @@ def visit(node):
         VariableTracker.apply(visit, (args, kwargs))
 
         if requires_grad and torch.is_grad_enabled():
-            # TODO(jansel): handle this in training mode
-            unimplemented("autograd.Function with requires_grad")
+            from .builder import VariableBuilder
+
+            def trampoline_autograd_fn(*args, **kwargs):
+                return self.fn_cls.apply(*args, **kwargs)
+
+            torch._dynamo.allow_in_graph(trampoline_autograd_fn)
```

Comment:
Any way we could get dynamo to trace self._fn_cls.forward to ensure the correctness even if the graph that is passed down contains this call directly?

Reply:
Yes, I have plans to ensure this is safe via export fwd and bwd and if it exports, installing the guards from it and proceeding. Still designing the plan :)